domain/secret/service/service.go

```diff
@@ -479,14 +467,33 @@ func (s *SecretService) UpdateCharmSecret(ctx context.Context, uri *secrets.URI,
 		}()
 	}
 
-	err := s.secretState.UpdateSecret(ctx, uri, p)
-	if errors.Is(err, secreterrors.SecretLabelAlreadyExists) {
-		return errors.Errorf("secret with label %q is already being used", *params.Label)
-	}
-	if err != nil {
-		return errors.Annotatef(err, "cannot update charm secret %q", uri.ID)
-	}
-	return nil
+	err := s.secretState.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		if err := s.canManage(ctx, uri, params.Accessor, params.LeaderToken); err != nil {
+			return errors.Trace(err)
+		}
+
+		rotatePolicy := domainsecret.MarshallRotatePolicy(params.RotatePolicy)
+		p.RotatePolicy = &rotatePolicy
+		if params.RotatePolicy.WillRotate() {
+			policy, err := s.secretState.GetRotatePolicy(ctx, uri)
+			if err != nil {
+				return fmt.Errorf("cannot get rotate policy for secret %q: %w", uri.ID, err)
+			}
+			if !policy.WillRotate() {
```

Comment:
We need to include a check when updating that the rotation policy is still "never" at the time of the update. Consider the case where this txn is retried and another one gets to update the rotate policy before this one is re-run. 
Usually the unit agent hook serialisation will prevent this UpdateCharmSecret method from being called concurrently, but there's a corner case where actions might be run in parallel and they could use juju-exec to update a secret. We might want to confirm this use case does result in concurrent calls.